spec/unit/file_system_spec.rb

```diff
@@ -872,4 +872,195 @@ def increment_counter_in_multiple_processes(file, num_procs, options)
+
+        expect {
+          Puppet::FileSystem.replace_file(dir) { |f| f.write(content) }
+        }.to raise_error(Errno::EISDIR, /Is a directory/)
+      end
+
+      it 'preserves the existing content if an error is raised' do
+        File.write(dest, 'existing content')
+
+        Puppet::FileSystem.replace_file(dest) { |f| raise 'whoops' } rescue nil
+
+        expect(Puppet::FileSystem.binread(dest)).to eq('existing content')
+      end
+
+      context 'on posix', unless: Puppet::Util::Platform.windows? do
+        it 'preserves the existing mode' do
+          Puppet::FileSystem.chmod(0600, dest)
+
+          Puppet::FileSystem.replace_file(dest) { |f| f.write(content) }
+
+          mode = Puppet::FileSystem.stat(dest).mode
+          expect(mode & 07777).to eq(0600)
+        end
+
+        it 'applies the specified mode' do
+          Puppet::FileSystem.chmod(0600, dest)
+
+          Puppet::FileSystem.replace_file(dest, 0777) { |f| f.write(content) }
+
+          mode = Puppet::FileSystem.stat(dest).mode
+          expect(mode & 07777).to eq(0777)
+        end
+
+        it 'updates a read-only file' do
+          Puppet::FileSystem.chmod(0400, dest)
+
+          Puppet::FileSystem.replace_file(dest) { |f| f.write(content) }
+
+          expect(Puppet::FileSystem.binread(dest)).to eq(content)
+
+          mode = Puppet::FileSystem.stat(dest).mode
+          expect(mode & 07777).to eq(0400)
+        end
+      end
+
+      context 'on windows', if: Puppet::Util::Platform.windows? do
+        it 'preserves the existing mode' do
+          old_sd = Puppet::Util::Windows::Security.get_security_descriptor(dest)
+
+          Puppet::FileSystem.replace_file(dest) { |f| f.write(content) }
+
+          new_sd = Puppet::Util::Windows::Security.get_security_descriptor(dest)
+          expect(old_sd.owner).to eq(new_sd.owner)
+          expect(old_sd.group).to eq(new_sd.group)
+          old_sd.dacl.each do |ace|
+            expect(new_sd.dacl).to include(an_object_having_attributes(sid: ace.sid, mask: ace.mask))
+          end
+        end
+
+        it 'applies the specified mode' do
+          Puppet::FileSystem.replace_file(dest, 0644) { |f| f.write(content) }
+
+          current_sid = Puppet::Util::Windows::SID.name_to_sid(Puppet::Util::Windows::ADSI::User.current_user_name)
+          sd = Puppet::Util::Windows::Security.get_security_descriptor(dest)
+          expect(sd.dacl).to contain_exactly(
+            an_object_having_attributes(sid: 'S-1-5-18', mask: 0x1f01ff),
+            an_object_having_attributes(sid: 'S-1-5-32-544', mask: 0x1f01ff),
+            an_object_having_attributes(sid: current_sid, mask: 0x1f01ff),
+            an_object_having_attributes(sid: 'S-1-5-32-545', mask: 0x120089)
+          )
+        end
+
+        it 'raises Errno::EACCES if access is denied' do
+          Puppet::Util::Windows::Security.stubs(:get_security_descriptor).raises(Puppet::Util::Windows::Error.new('access denied', 5))
```

Comment:
After these changes were merged to master, Jacob switched puppet from mocha to rspec for mocking and stubbing. So you'll want to add a commit that changes it to use rspec:

```ruby
allow(Puppet::Util::Windows::Security).to receive(:get_security_descriptor).and_raise(Puppet::Util::Windows::Error.new('access denied', 5))
```

Reply:
Yes - just looking through @jhelwig at the moment.

Reply:
So - that is Commit c4b0f889a57cb06d758ec4cca896ab87d46f714d  
Gives the following cherry-pick error/status:
```
You are currently cherry-picking commit c4b0f889a5.
  (fix conflicts and run "git cherry-pick --continue")
  (use "git cherry-pick --abort" to cancel the cherry-pick operation)

Unmerged paths:
  (use "git add <file>..." to mark resolution)

        both modified:   spec/unit/application_spec.rb
        both modified:   spec/unit/configurer_spec.rb
        both modified:   spec/unit/network/http/connection_spec.rb
        both modified:   spec/unit/provider/package/dpkg_spec.rb
        both modified:   spec/unit/provider/package/gem_spec.rb
        both modified:   spec/unit/provider/package/pip_spec.rb
        both modified:   spec/unit/provider/package/puppet_gem_spec.rb
        both modified:   spec/unit/provider/package/rpm_spec.rb
        both modified:   spec/unit/provider/service/upstart_spec.rb
        both modified:   spec/unit/provider/user/directoryservice_spec.rb
        both modified:   spec/unit/ssl/validator_spec.rb
        both modified:   spec/unit/transaction/event_manager_spec.rb
        both modified:   spec/unit/util/pidlock_spec.rb

no changes added to commit (use "git add" and/or "git commit -a")
```

Reply:
You shouldn't need to cherry-pick any of the mocha removal commits. It looks like this PR is based on a version of 5.5.x that already contains the changes.

Reply:
Tx @jhelwig Maybe this is do-able in 5.5.x then.

Reply:
done